source/Patches/Roles/Role.cs

```diff
@@ -15,23 +17,39 @@
 
 namespace TownOfUs.Roles
 {
-    public abstract class Role
+    public abstract class Role : IRoleDetails
     {
+        private static List<Role> _roleSingletons { get; } = new List<Role>();
+        public static ReadOnlyCollection<Role> RoleSingleton => _roleSingletons.AsReadOnly();
+
+        public static void AddSingleton(Role role)
```

Comment:
If we’re going to do this, let’s make it a RoleEnum:Role (or RoleEnum:RoleSingleton) dictionary. That will save all that pointless iterating through the whole list to find the relevant role.

Reply:
Sounds good.

Reply:
> > I am nervous about all these Role subclasses now also having no-arg constructors. Makes me worry that it will be possible to now have an object of type `Seer` that doesn’t have a player attached. Seems easy to use the wrong one accidentally. I think this is fixable by having the relevant methods return `RoleSingleton` instead?
> 
> RoleSingleton creates an object through parameter-less constructor while `Role.Gen<T>` uses the one with a single `PlayerControl` parameter. Parameter-less constructor wasn't there before, so I'm very sure it won't affect the previous code.
> 
> > Seems easy to use the wrong one accidentally
> 
> We can use [System.ObsoleteAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-6.0) if you liked to. It will throw a warning/error to the compiler if people attempted to use it (unless they are also using `Activator.CreateInstance`).

I am not worried about breaking code that's already there. I'm worried about what happens when people who don't understand what you did or why coming along and using the wrong one in the future. I believe you can return `RoleSingleton` instead of `Role`, which will help.

I did not know about the `ObsoleteAttribute`. That's very nice. If we make that an error, that will probably help a lot.